### DISCUSSION TITLE
Keycloak needs a CSS injection textbox in the dashboard

### DISCUSSION THREAD

**Discussion by iyobo** [keycloak/keycloak#25480] [Ideas]
I find that the main issue with Keycloak is the lack of "easy" theme customization.
There should be a UI to customize what the login pages should look like from a minimalistic perspective, per realm....as opposed to having to compile a jar and deal with freemarker.

This isn't asking for much. The theme customizer UI should simply provide a textbox where the user can put in some CSS to override everything e.g background image, logo and other CSS styling. This css should then be injected into the default theme template.

**Comment by jonkoops** [comment]
You don't need to compile a JAR, if you want to create a custom theme with CSS you can do so by placing the assets in the `themes` directory (see [themes documentation](https://www.keycloak.org/docs/latest/server_development/index.html#creating-a-theme)). You can then configure that theme on a per realm basis.

**Comment by iyobo** [comment]
@jonkoops But there-in lies the issue. My proposal is about being able to soft customize themes without touching folders, files, etc in the source code.

Many use keycloak simply by deploying the docker image into a cluster (portainer, kubernetes, etc), and do not have to touch source code. The "hard structure" of the default keycloak theme is actually very good, but allowing for "soft CSS" customizations through CSS injection defined per realm from the dashboard would be stellar. No actual files or folders need be touched/added to the deployment to augment the look per realm; No restarts to re-load those files; etc. These are the modern benefits of what I'm proposing.

This feature would bring keycloak on par with the other modern open-source IDPs out there today that allow for this kind of soft dynamic look-and-feel change, but with the same "hard theme".

**Reply by jonkoops** [reply]
The directory I mentioned is not in the source code, but part of the Keycloak distribution itself. If you are using containers you can mount your customizations on top of that directory.

I personally don't see the need to add yet another global realm setting for something that can easily be customized with the existing theme infrastructure. Allowing users to specify styles that can be injected directly into a page also has implications for Content Security Policy.

Feel free to log this as a feature request in the issue tracker to see if others might find this useful, if it gets enough thumbs up we might consider it for implementation.
